### ISSUE TITLE
Get logits/predictions from a detector instead of loss

### ISSUE THREAD

**Issue by CMobley7** [open-mmlab/mmdetection#3933]
Is there a way to load a detector, feed it an img and img_meta, and retrieve the class logits/predictions instead of the loss? It looks like the default behavior of forward_train and forward_test is to return the loss. After looking through the code, it looks like forward_dummy, https://github.com/open-mmlab/mmdetection/blob/547eb8d12c1b29ad9644160ca5dcb223cc1d0ce6/mmdet/models/roi_heads/standard_roi_head.py,  might give me what I need. Is this correct or is there a different way to achieve my goal? Thanks in advance!

**Comment by CMobley7** [comment]
Upon a deeper dive of forward_dummy, it look like it's generating random proposals; so, the bbox logits/predictions wouldn't be usable. Is there an easy way to return this information?

**Comment by CMobley7** [comment]
@hellock  and @xvjiarui,
The following code gets me the `logits`, as well as the `bboxes`. However, even though the model I'm using only has 12 classes, which I verified with `print(model.CLASSES)` and `print(len(model.CLASSES))`, I get a `1000x320` tensor for `bboxes` and a `1000x81` tensor for the `logits`. It looks like it defaults to the number of COCO classes. How can I override this to be the number of classes in `model.CLASSES`? Also, for the `logits`, is the first element of each row, the `background class`? 
```
# build the model from a config file and a checkpoint file
model = init_detector(config=config, checkpoint=checkpoint, device=device)


img_metas = {}
channels, height, width = img.shape[-3:]
img_shape = (int(height), int(width), int(channels))
img_metas["img_shape"] = img_shape
img_metas["scale_factor"] = 1.0
    
img = torch.from_numpy(img).to(device).float()

with torch.no_grad():
    x = model.extract_feat(img)
    proposal_list = model.rpn_head.simple_test_rpn(x, img_metas=[img_metas])
    rois = bbox2roi(proposal_list)
    bbox_results = model.roi_head._bbox_forward(x, rois)
    logits = bbox_results["cls_score"]
    bbox_pred = bbox_results["bbox_pred"]
    bboxes = model.roi_head.bbox_head.bbox_coder.decode(rois[:, 1:], bbox_pred, max_shape=img_shape[:2])
```

**Comment by xvjiarui** [comment]
Hi @CMobley7 
You may check the `num_classes` of heads.

**Comment by CMobley7** [comment]
@xvjiarui, you're right. It looks like `num_classes` is hardcoded in most configs for the `roi_head`, `mask_head`, `semantic_head` and `mask_iou_head`. I see now that https://mmdetection.readthedocs.io/en/latest/tutorials/finetune.html, instructs one to change this variable, my bad. I was under the false impression that this would be set dynamically based off of one's dataset, i.e. the classes var in the config ( https://mmdetection.readthedocs.io/en/latest/tutorials/customize_dataset.html). Is there anyway, `num_classes` can be added dynamically via doing a `len` of the `classes` var? 

Also, it appears that when I use `tools/test.py`, it outputs the right number of classes when doing evaluation and inference. However, when I examine the code, it looks like it should be outputting 80 classes based off of num_classes in the particular head, not the number of classes in my custom dataset. How is this possible?

**Comment by LiUzHiAn** [comment]
Hi, I wrote a function that can return the class probabilities for each detection.
```
def inference_detector_with_probs(model, img):
    """Inference image with the detector.
    Args:
        model (nn.Module): The loaded detector.
        img (str): image file.
    Returns:
    """

    cfg = model.cfg
    device = next(model.parameters()).device  # model device
    cfg.data.test.pipeline = replace_ImageToTensor(cfg.data.test.pipeline)
    test_pipeline = Compose(cfg.data.test.pipeline)

    datas = []
    # add information into dict
    data = dict(img_info=dict(filename=img), img_prefix=None)
    # build the data pipeline
    datas.append(test_pipeline(data))

    data = collate(datas, samples_per_gpu=len(imgs))
    # just get the actual data from DataContainer
    data['img_metas'] = [img_metas.data[0] for img_metas in data['img_metas']]
    data['img'] = [img.data[0] for img in data['img']]

    # scatter to specified GPU
    data = scatter(data, [device])[0]

    img_metas = data["img_metas"][0]
    # forward the model
    with torch.no_grad():
        x = model.extract_feat(data['img'][0])
        proposal_list = model.rpn_head.simple_test_rpn(x, img_metas)

        img_shapes = tuple(meta['img_shape'] for meta in img_metas)
        scale_factors = tuple(meta['scale_factor'] for meta in img_metas)

        rois = torch.stack(proposal_list, dim=0)

        batch_index = torch.arange(
            rois.size(0), device=rois.device).float().view(-1, 1, 1).expand(
            rois.size(0), rois.size(1), 1)
        rois = torch.cat([batch_index, rois[..., :4]], dim=-1)
        batch_size = rois.shape[0]
        num_proposals_per_img = rois.shape[1]

        # Eliminate the batch dimension
        rois = rois.view(-1, 5)
        bbox_results = model.roi_head._bbox_forward(x, rois)
        cls_logits = bbox_results['cls_score']
        bbox_pred = bbox_results['bbox_pred']

        # Recover the batch dimension
        rois = rois.reshape(batch_size, num_proposals_per_img, -1)
	# logits
        cls_logits = cls_logits.reshape(batch_size, num_proposals_per_img, -1)
        bbox_pred = bbox_pred.reshape(batch_size, num_proposals_per_img, -1)

        scores = F.softmax(cls_logits, dim=-1)
        bboxes = model.roi_head.bbox_head.bbox_coder.decode(rois[..., 1:], bbox_pred, max_shape=img_shapes)
        # B, 1, bboxes.size(-1)
        scale_factor = bboxes.new_tensor(scale_factors).unsqueeze(1).repeat(
            1, 1,
            bboxes.size(-1) // 4)
        bboxes /= scale_factor

        det_bboxes = []
        det_labels = []
        det_probs = []
        for (bbox, score) in zip(bboxes, scores):
            det_bbox, det_label, det_prob = my_multiclass_nms(bbox, score,
                                                              cfg.model.test_cfg.rcnn.score_thr,
                                                              cfg.model.test_cfg.rcnn.nms,
                                                              cfg.model.test_cfg.rcnn.max_per_img)

            det_bboxes.append(det_bbox)
            det_labels.append(det_label)
            det_probs.append(det_prob)

    return det_bboxes, det_labels, det_probs
```
For more details and usage example, please check out from [>>my gist here<<](https://gist.github.com/LiUzHiAn/13f8f13b354e848a3b166fde3cc3335a). Hope it helps!
